contracts/src/Consensus.sol

```diff
@@ -0,0 +1,460 @@
+    EXTERNAL FUNCTIONS - EPOCH MANAGEMENT
+    ============================================================
+
+    3. incrementEpoch()
+       - Initiates KeyGen for a new epoch
+       - Can be called by anyone
+       - Takes no parameters
+       - Enforces 86400 second (1 day) minimum between epochs
+
+       Validations:
+       - block.timestamp >= lastEpochTimestamp + 86400
+       - validatorsList.length >= threshold (sufficient validators registered)
+
+       Actions:
+       - Increment currentEpoch
+       - Set lastEpochTimestamp = block.timestamp
+       - Call _initiateKeyGen()
+
+    4. retryKeyGen()
+       - Retries KeyGen for current epoch if it failed - @CHECK How to detect failure?
+       - Can be called by anyone
+       - Useful if KeyGen ceremony fails and needs to be restarted
+
+       Actions:
+       - Call FROSTCoordinator.keygenAbort(currentEpoch) first (to clean up failed state)
+       - Call _initiateKeyGen() to restart
+       - Emit KeyGenRetried(currentEpoch, participantsMerkleRoot)
+
+    ============================================================
+    INTERNAL FUNCTIONS - KEYGEN
+    ============================================================
+
+    5. _initiateKeyGen() internal
+       - Internal function to initiate KeyGen ceremony
+       - Called by constructor and incrementEpoch()
+
+       Actions:
+       - Calculate merkle root from sorted validatorsList - @CHECK Should we use FROSTMerkleMap.init() or something else?
+       - Store merkle root: epochParticipants[currentEpoch] = merkleRoot
+       - Store count: epochParticipantCount[currentEpoch] = validatorsList.length
+       - Call FROSTCoordinator.keygen(
+             uint96(currentEpoch),               // nonce = epoch number
+             merkleRoot,                         // participants merkle root
+             uint128(validatorsList.length),     // count (auto-calculated)
+             threshold                           // current threshold value
+         )
+       - Emit EpochIncremented(currentEpoch, merkleRoot, count, threshold, block.timestamp)
+       - Keep registeredValidators mapping (validators stay registered unless they deregister)
+
+    ============================================================
+    EXTERNAL FUNCTIONS - TRANSACTION PROPOSALS
+    ============================================================
+
+    6. proposeSafeTransaction(
+           address safeAddress,
+           address to,
+           uint256 value,
+           bytes calldata data,
+           uint8 operation,
+           uint256 safeTxGas,
+           uint256 baseGas,
+           uint256 gasPrice,
+           address gasToken,
+           address refundReceiver,
+           uint256 nonce,
+           uint256 chainId
+       ) external returns (bytes32 txHash)
+
+       - Allows anyone to propose a Safe transaction for attestation
+       - Stores full transaction with current epoch
```

Comment:
We do not necessarily need to store the information onchain, emitting an event would do and only store minimal information onchain (i.e. the block number for easier log retrieval, epoch number, ...). 

For reference: this approach is used in Secret Harbour